integration_tests/conftest.py

```diff
@@ -272,51 +272,145 @@ def _calculate_bayes_factor(
+                    # mean_lower_bound = mean
+                    mean_min = mean
+                    mean_bounds_changed = True
+                elif lb_cdf <= ub_cdf < (1 - 0.025) and 0 <= lb_cdf < 0.025:
+                    # The distribution is low on both quantiles, so we need to increase the mean
+                    # mean_upper_bound = mean
+                    mean_max = mean
+                    mean_bounds_changed = True
+
+                if not concentration_bounds_changed and not mean_bounds_changed:
+                    break
+
+                if concentration_bounds_changed:
+                    # We have been optimizing mean with inaccurate concentration bounds; let's restart
+                    # our mean search (which is pretty small/cheap).
+                    mean_max = upper_bound
+                    mean_min = lower_bound
+
+                if mean_bounds_changed:
+                    mean = (mean_min + mean_max) / 2
+                    # We have been optimizing concentration with inaccurate mean bounds; let's back off
+                    # a bit to explore concentration more.
+                    # NOTE: The convergence of this method depends pretty crucially on this backoff
+                    # constant. Without it, we don't converge at all in some cases.
+                    # If it is too high, convergence is slow and sometimes runs out of iterations.
+                    # 1.75 worked well across a wide range of inputs in preliminary testing.
+                    concentration_max = min(concentration_max * 1.75, 1e40)
+                    concentration_min = max(concentration_min / 1.75, 1e-3)
 
-        assert optimization_result.success or optimization_result.fun < 1e-05
+        assert (
+            best_error < 0.1
+        ), f"Beta distribution fitting for {lower_bound}, {upper_bound} failed with UI squared error {best_error}"
+        if best_error > 1e-5:
+            warnings.warn(
+                f"Didn't find a very good beta distribution for {lower_bound}, {upper_bound} -- using a best guess with UI squared error {best_error}"
+            )
 
-        result = np.exp(optimization_result.x)
+        result = (
+            best_mean * best_concentration,
+            (1 - best_mean) * best_concentration,
+        )
         assert len(result) == 2
         return tuple(result)
 
+    def _ui_squared_error(
+        self, concentration: float, mean: float, lower_bound: float, upper_bound: float
+    ) -> float:
+        a = mean * concentration
+        b = (1 - mean) * concentration
+        dist = scipy.stats.beta(a=a, b=b)
+
+        squared_error_lower = self._quantile_squared_error(dist, lower_bound, 0.025)
+        squared_error_upper = self._quantile_squared_error(dist, upper_bound, 0.975)
+
+        try:
+            return squared_error_lower + squared_error_upper
+        except FloatingPointError:
```

Comment:
Does this essentially raise this error before we would expect to hit it how we were with the current failures due to hardware?

Reply:
This catches errors due to error values that are too large to be stored in a float. Overall, this PR changes what we were doing completely!